File: ConsumerVPN-Android/consumerVPN/src/main/java/com/wlvpn/slider/whitelabelvpn/jobs/TokenRefreshJob.java
package com.wlvpn.slider.whitelabelvpn.jobs;

import android.support.annotation.NonNull;

import com.evernote.android.job.Job;
import com.evernote.android.job.JobRequest;
import com.gentlebreeze.vpn.http.api.error.LoginErrorThrowable;
import com.wlvpn.slider.whitelabelvpn.BuildConfig;
import com.wlvpn.slider.whitelabelvpn.ConsumerVpnApplication;
import com.wlvpn.slider.whitelabelvpn.auth.Credentials;
import com.wlvpn.slider.whitelabelvpn.auth.CredentialsManager;
import com.wlvpn.slider.whitelabelvpn.managers.AccountManager;
import com.wlvpn.slider.whitelabelvpn.managers.ConnectableManager;
import com.wlvpn.slider.whitelabelvpn.managers.SettingsManager;
import com.wlvpn.slider.whitelabelvpn.models.ApiErrorCodes;

import kotlin.Unit;
import timber.log.Timber;

/**
 * Job implementation for updating the server list.
 */
public class TokenRefreshJob extends Job {

    /**
     * Job ID
     */
    static final String JOB_ID = "TokenRefreshJob";

    private final CredentialsManager credentialsManager;
    private final AccountManager accountManager;
    private final SettingsManager settingsManager;
    private final ConnectableManager connectableManager;

    TokenRefreshJob(CredentialsManager credentialsManager,
                    AccountManager accountManager,
                    SettingsManager settingsManager,
                    ConnectableManager connectableManager) {
        this.credentialsManager = credentialsManager;
        this.accountManager = accountManager;
        this.settingsManager = settingsManager;
        this.connectableManager = connectableManager;
    }

    /**
     * Schedules the job.
     */
    public static void schedule() {
        //For debug builds use reduced timers and more accepting connection type
        long start = BuildConfig.DEBUG ?
                ConsumerVpnJobCreator.FIFTEEN_MINUTE_IN_MILLISECONDS
                : ConsumerVpnJobCreator.TWENTY_FOUR_HOURS_IN_MILLISECONDS;

        JobRequest.NetworkType networkType = BuildConfig.DEBUG ?
                JobRequest.NetworkType.CONNECTED
                : JobRequest.NetworkType.UNMETERED;

        new JobRequest.Builder(JOB_ID)
                .setRequiredNetworkType(networkType)
                .setRequirementsEnforced(true)
                .setPeriodic(start)
                .setUpdateCurrent(true)
                .build()
                .schedule();
    }

    @NonNull
    @Override
    protected Result onRunJob(@NonNull Params params) {

        if (accountManager.isUserLoggedIn()
                && !ConsumerVpnApplication.getVpnSdk().isAccessTokenValid()) {

            // Blocking current thread to allow the refresh token refresh finish
            // before sending Result.Success to Job
            ConsumerVpnApplication.getVpnSdk()
                    .refreshToken()
                    .blockCurrentThread()
                    .subscribe(
                            vpnLoginResponse -> Unit.INSTANCE,
                            throwable -> {
                                Timber.e(throwable, "Token refresh failed during job");
                                tokenFailLogin();
                                return Unit.INSTANCE;
                            });
        }

        return Result.SUCCESS;
    }

    /**
     * Will update db values and will logout user whenever credentials changed on server
     */
    private void tokenFailLogin() {
        Credentials credentials = credentialsManager.getCredentials();

        // Blocking current thread to allow the login finish
        // before sending Result.Success to Job
        ConsumerVpnApplication.getVpnSdk()
                .loginWithUsername(
                        credentials.getUsername(),
                        credentials.getPassword())
                .blockCurrentThread()
                .subscribe(
                        vpnLoginResponse -> Unit.INSTANCE,
                        throwable -> {
                            if (throwable instanceof LoginErrorThrowable) {
                                LoginErrorThrowable loginThrowable =
                                        (LoginErrorThrowable) throwable;
                                // When login fails that means user updated
                                // it's credentials on web the user will be logged out
                                if (loginThrowable.getResponseCode() ==
                                        ApiErrorCodes.INVALID_CREDENTIALS) {

                                    if (ConsumerVpnApplication.getVpnSdk().isConnected()) {
                                        ConsumerVpnApplication.getVpnSdk()
                                                .disconnect();
                                    }
                                    accountManager.logout();
                                    settingsManager.clear();
                                    connectableManager.clear();

                                    ConsumerVpnApplication.getVpnSdk().logout();
                                }
                            }
                            return Unit.INSTANCE;
                        });
    }
}

File: ConsumerVPN/app/src/main/java/com/wlvpn/consumervpn/data/job/TokenRefreshJob.kt
package com.wlvpn.consumervpn.data.job

import com.evernote.android.job.Job
import com.evernote.android.job.JobRequest
import com.wlvpn.consumervpn.BuildConfig
import com.wlvpn.consumervpn.data.failure.NetworkNotAvailableFailure
import com.wlvpn.consumervpn.data.gateway.authorization.failure.RefreshTokenReLoginFailure
import com.wlvpn.consumervpn.domain.service.authentication.UserAuthenticationService
import com.wlvpn.consumervpn.domain.service.authorization.UserAuthorizationService
import com.wlvpn.consumervpn.domain.service.authorization.failure.UserNotAuthorizedFailure
import com.wlvpn.consumervpn.domain.service.vpn.VpnService
import timber.log.Timber

/**
 * Job implementation for refreshing the session token
 */
class TokenRefreshJob internal constructor(
    private val userAuthenticationService: UserAuthenticationService,
    private val userAuthorizationService: UserAuthorizationService,
    private val vpnService: VpnService
) : Job() {

    companion object {

        internal const val JOB_ID = "consumer:TokenRefreshJob"

        fun schedule() {
            //For debug builds use reduced timers and more accepting connection listType
            val start =
                if (BuildConfig.DEBUG) FIFTEEN_MINUTES_IN_MILLISECONDS
                else TWENTY_FOUR_HOURS_IN_MILLISECONDS

            JobRequest.Builder(JOB_ID)
                .setRequiredNetworkType(JobRequest.NetworkType.CONNECTED)
                .setRequirementsEnforced(true)
                .setPeriodic(start)
                .setUpdateCurrent(true)
                .build()
                .schedule()
        }
    }

    override fun onRunJob(params: Params): Result {
        userAuthorizationService.refreshToken()
            .blockingGet()
            ?.let { throwable ->
                when (throwable) {
                    is UserNotAuthorizedFailure ->
                        Timber.e(throwable, "Impossible to refresh token. The user is not authorized")

                    is RefreshTokenReLoginFailure -> {
                        Timber.e(throwable, "Re-login failed while refreshing token. Login out user")

                        // Logout clears the background jobs
                        vpnService.disconnect()
                            .onErrorComplete()
                            .andThen(userAuthenticationService.logout())
                            .blockingGet()
                            .let { logoutThrowable ->
                                Timber.e(logoutThrowable, "Error while login out user on token refresh fail")
                            }
                    }

                    is NetworkNotAvailableFailure ->
                        Timber.e(throwable, "Network was not available while refreshing token")

                    else -> Timber.e(throwable, "An error occurred while refreshing token")
                }
            }

        return Result.SUCCESS
    }
}
